components/vocabularies/oncotree/api/src/main/java/org/phenotips/vocabulary/internal/solr/OncoTree.java

```diff
@@ -0,0 +1,322 @@
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Provides access to the OncoTree vocabulary. The vocabulary prefix is {@code ONCO}.
+ *
+ * @version $Id$
+ * @since 1.4
+ */
+@Component
+@Named("onco")
+@Singleton
+public class OncoTree extends AbstractCSVSolrVocabulary
+{
+    /** The default location of the OncoTree data file. */
+    private static final String SOURCE = "http://oncotree.mskcc.org/oncotree/api/tumor_types.txt";
+
+    /** A picker for {@link OncoTreePropertyWriteStrategy} objects. */
+    private static final OncoTreePropertyWriteStrategyPicker STRATEGIES = new OncoTreePropertyWriteStrategyPicker();
+
+    /** Column labels for the tab-separated cancers file. */
+    private static final String[] COLUMNS = {
+        OncoTreeUtils.TISSUE,
+        OncoTreeUtils.CANCER + "1",
+        OncoTreeUtils.CANCER + "2",
+        OncoTreeUtils.CANCER + "3",
+        OncoTreeUtils.CANCER + "4",
+        OncoTreeUtils.TERM_GROUP,
+        OncoTreeUtils.COLOUR,
+        OncoTreeUtils.NCI_ID,
+        OncoTreeUtils.UMLS_ID
+    };
+
+    @Override
+    protected int getSolrDocsPerBatch()
+    {
+        return 15000;
+    }
+
+    @Override
+    protected Collection<SolrInputDocument> load(@Nonnull final URL url)
+    {
+        final Collection<SolrInputDocument> data = new ArrayList<>();
+        try (final BufferedReader in = new BufferedReader(
+            new InputStreamReader(getInputStream(url), StandardCharsets.UTF_8))) {
+            final CSVFormat parser = setupCSVParser();
+            // Process each csv record row.
+            for (final CSVRecord row : parser.parse(in)) {
+                processDataRow(row, data);
+            }
+        } catch (final IOException e) {
+            this.logger.error("Failed to load vocabulary source: {}", e.getMessage());
+        }
+        return data;
+    }
+
+    /**
+     * Gets an input stream from the provided {@code url}.
+     *
+     * @param url the {@link URL} for the cancers data
+     * @return an {@link InputStream} with the data
+     * @throws IOException if a connection cannot be opened
+     */
+    @VisibleForTesting
+    InputStream getInputStream(@Nonnull final URL url) throws IOException
+    {
+        return url.openConnection().getInputStream();
+    }
+
+    /**
+     * Processes a CSV {@code row}, and adds the relevant fields to a collection of {@code data}.
+     *
+     * @param row a {@link CSVRecord} containing data for one path (from root to leaf) of the OncoTree
+     * @param data a collection of {@link SolrInputDocument} objects where data from {@code row} will be written
+     */
+    private void processDataRow(@Nonnull final CSVRecord row, @Nonnull final Collection<SolrInputDocument> data)
+    {
+        final SolrInputDocument doc = new SolrInputDocument();
+
+        for (int i = 0; i < row.size(); i++) {
+            final String label = COLUMNS[i];
+            final String value = row.get(i);
```

Comment:
I feel that 4 files with WriteStrategies of total ~400 lines is overcomplication of what can be done in 30 following lines.
` if (i == 0) {`
       
       String tissueName = StringUtils.substringBefore(value, OPEN_BRACKET).trim();

	if (StringUtils.isNotBlank(tissueName)) {

        doc.addField(COLUMNS[0], tissueName);

    }
`} else if (i < 5 ) {`

    // process cancer name

    String prevId = (String) doc.getFieldValue(OncoTreeUtils.ID);

    if (StringUtils.isNotBlank(prevId)) {

        doc.setField(IS_A, prevId);

        doc.addField(TERM_CATEGORY, prevId);

    }

	String id = StringUtils.substringBetween(name, OPEN_BRACKET, CLOSED_BRACKET);

    if (StringUtils.isNotBlank(id)) {

        doc.setField(ID, ID_PREFIX + id);
    } else {

        this.logger.warn("No identifier could be extracted from the provided cancer name: {}", name);

        doc.setField(ID, ID_PREFIX + name);

    }

	String prevId = (String) doc.getFieldValue(OncoTreeUtils.ID);

    if (StringUtils.isNotBlank(prevId)) {

        doc.setField(IS_A, prevId);

        doc.addField(TERM_CATEGORY, prevId);

    }

`} else {`

    doc.addField(COLUMNS[i], value);

`}`